### ISSUE TITLE
no-constant-condition doesn't respect generator function expressions

### ISSUE THREAD

**Issue by helixbass** [eslint/eslint#10826]
**Tell us about your environment**
Came across this while working on [`eslint-plugin-coffee`](https://github.com/helixbass/eslint-plugin-coffee) custom parser
* **ESLint Version:**
5.4.0
* **Node Version:**
8.11.4
* **npm Version:**
6.2.0

**What parser (default, Babel-ESLint, etc.) are you using?**
Coffeescript, but the issue seems to apply regardless of parser

**Please show your full configuration:**
The relevant config would be `no-constant-condition: error`

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**
I was attempting to run equivalent Coffeescript versions of the tests in `tests/lib/rules/no-constant-condition.js` against the `no-constant-condition` rule. I saw tests fail eg this was expected to throw an error but didn't:

```
while true
  foo = -> yield
```
Which would be basically equivalent in JS to:
```
while (true) {
  const foo = function*() { yield }
}
```

**What did you expect to happen?**
For the test to pass (the nested `yield` shouldn't suppress the warning for the constant `while` condition)

**What actually happened? Please include the actual, raw output from ESLint.**
The test failed (no warning was reported by `no-constant-condition`)

Looking at the source code for `lib/rules/no-constant-condition.js`, I saw that it only "resets the `yield` checker" when entering a `FunctionDeclaration` (but not a `FunctionExpression`). Coffeescript doesn't generate `FunctionDeclaration`s so that's why I ran into this

It seems to me that generator `FunctionExpression`s should behave the same way as generator `FunctionDeclaration`s

Will submit a fixing PR momentarily

**Comment by not-an-aardvark** [comment]
Closing because this was fixed by https://github.com/eslint/eslint/pull/10827.
